File: sdks/python/src/echo_agent_sdk/retry_policy_values.py
from __future__ import annotations

import random
from dataclasses import dataclass, replace


@dataclass(frozen=True)
class RetryPolicy:
    max_retries: int
    base_delay_ms: float
    max_delay_ms: float = 60_000
    jitter_enabled: bool = False

    @classmethod
    def new(cls, max_retries: int, base_delay_ms: float) -> RetryPolicy:
        if max_retries < 0 or base_delay_ms < 0:
            raise ValueError("retry policy values must be non-negative")
        return cls(max_retries, base_delay_ms)

    @classmethod
    def default(cls) -> RetryPolicy:
        return cls(3, 500, 30_000, True)

    @classmethod
    def no_retry(cls) -> RetryPolicy:
        return cls(0, 0, 0, False)

    def max_delay(self, delay_ms: float) -> RetryPolicy:
        if delay_ms < 0:
            raise ValueError("retry max delay must be non-negative")
        return replace(self, max_delay_ms=delay_ms)

    def jitter(self, enabled: bool) -> RetryPolicy:
        return replace(self, jitter_enabled=enabled)

    def delay_for(self, attempt: int) -> float:
        if attempt < 0:
            raise ValueError("retry attempt must be non-negative")
        if attempt == 0:
            return 0
        capped = min(
            self.base_delay_ms * (2 ** min(attempt - 1, 10)), self.max_delay_ms
        )
        return (
            random.uniform(0, capped) if self.jitter_enabled and capped > 0 else capped
        )

File: sdks/typescript/src/retry_policy_values.ts
/** Retry policy configuration and deterministic backoff projection. */
export class RetryPolicy {
  private constructor(
    readonly maxRetries: number,
    readonly baseDelayMs: number,
    readonly maxDelayMs: number,
    readonly jitterEnabled: boolean,
  ) {}

  static new(maxRetries: number, baseDelayMs: number): RetryPolicy {
    if (!Number.isSafeInteger(maxRetries) || maxRetries < 0 || !Number.isFinite(baseDelayMs) || baseDelayMs < 0) {
      throw new RangeError("retry policy values must be non-negative");
    }
    return new RetryPolicy(maxRetries, baseDelayMs, 60_000, false);
  }

  static default(): RetryPolicy { return new RetryPolicy(3, 500, 30_000, true); }
  static noRetry(): RetryPolicy { return new RetryPolicy(0, 0, 0, false); }
  maxDelay(delayMs: number): RetryPolicy {
    if (!Number.isFinite(delayMs) || delayMs < 0) throw new RangeError("retry max delay must be non-negative");
    return new RetryPolicy(this.maxRetries, this.baseDelayMs, delayMs, this.jitterEnabled);
  }
  jitter(enabled: boolean): RetryPolicy { return new RetryPolicy(this.maxRetries, this.baseDelayMs, this.maxDelayMs, enabled); }
  delayFor(attempt: number): number {
    if (!Number.isSafeInteger(attempt) || attempt < 0) throw new RangeError("retry attempt must be non-negative");
    if (attempt === 0) return 0;
    const exponent = Math.min(attempt - 1, 10);
    const capped = Math.min(this.baseDelayMs * 2 ** exponent, this.maxDelayMs);
    return this.jitterEnabled && capped > 0 ? Math.floor(Math.random() * (capped + 1)) : capped;
  }
}
